src/components/tl/ucp/alltoall/alltoall_onesided.c

```diff
@@ -11,55 +11,362 @@
+            return;
+        }
+    }
+
+    alltoall_onesided_wait_completion(task, npolls);
+    if (UCC_TL_UCP_TASK_ONESIDED_P2P_COMPLETE(task)) {
+        task->super.status = UCC_OK;
+    }
+out:
+    return;
+}
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_get_start(ucc_coll_task_t *ctask)
 {
-    ucc_tl_ucp_task_t *task     = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
-    ucc_tl_ucp_team_t *team     = TASK_TEAM(task);
-    ptrdiff_t          src      = (ptrdiff_t)TASK_ARGS(task).src.info.buffer;
-    ptrdiff_t          dest     = (ptrdiff_t)TASK_ARGS(task).dst.info.buffer;
-    size_t             nelems   = TASK_ARGS(task).src.info.count;
-    ucc_rank_t         grank    = UCC_TL_TEAM_RANK(team);
-    ucc_rank_t         gsize    = UCC_TL_TEAM_SIZE(team);
-    ucc_rank_t         start    = (grank + 1) % gsize;
-    long              *pSync    = TASK_ARGS(task).global_work_buffer;
-    ucc_mem_map_mem_h  src_memh = TASK_ARGS(task).src_memh.local_memh;
-    ucc_mem_map_mem_h *dst_memh = TASK_ARGS(task).dst_memh.global_memh;
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ptrdiff_t          src       = (ptrdiff_t)TASK_ARGS(task).src.info.buffer;
+    ptrdiff_t          dest      = (ptrdiff_t)TASK_ARGS(task).dst.info.buffer;
+    ucc_rank_t         grank     = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         gsize     = UCC_TL_TEAM_SIZE(team);
+    size_t             nelems    = TASK_ARGS(task).src.info.count;
+    size_t             ntokens   = task->alltoall_onesided.tokens;
+    size_t             npolls    = task->alltoall_onesided.npolls;
+    int64_t            cpolls    = 0;
+    ucc_mem_map_mem_h  src_memh  = TASK_ARGS(task).dst_memh.local_memh;
+    ucc_mem_map_mem_h *dst_memh  = TASK_ARGS(task).src_memh.global_memh;
+    uint32_t          *posted    = &task->onesided.get_posted;
+    uint32_t          *completed = &task->onesided.get_completed;
     ucc_rank_t         peer;
 
-    if (TASK_ARGS(task).flags & UCC_COLL_ARGS_FLAG_SRC_MEMH_GLOBAL) {
-        src_memh = TASK_ARGS(task).src_memh.global_memh[grank];
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    nelems = (nelems / gsize) * ucc_dt_size(TASK_ARGS(task).src.info.datatype);
+    for (peer = (grank + 1) % gsize; *posted < gsize;
+         peer = (peer + 1) % gsize) {
+        UCPCHECK_GOTO(ucc_tl_ucp_get_nb(PTR_OFFSET(dest, peer * nelems),
+                                        (void *)PTR_OFFSET(src, grank * nelems),
+                                        nelems, peer, src_memh, dst_memh, team,
+                                        task),
+                      task, out);
+
+        if (!alltoall_onesided_handle_completion(task, posted, completed,
+                                                 ntokens, npolls, peer,
+                                                 &cpolls)) {
+            break;
+        }
     }
```

Comment:
get_start and get_progress (same for put) are almost identical. Can you pls keep the logic in progress function only